File: backend/app/services/grok_service.py
import requests
import json
from typing import Dict, List, Optional, Any
import logging
from datetime import datetime
from ..core.config import settings

logger = logging.getLogger(__name__)

class GrokService:
    """Service for interacting with xAI Grok API"""
    
    def __init__(self):
        self.api_key = settings.grok_api_key
        self.api_base = settings.grok_api_base
        self.model = settings.grok_model
        self.max_tokens = settings.grok_max_tokens
        self.temperature = settings.grok_temperature
    
    def _make_request(self, endpoint: str, data: Dict[str, Any]) -> Dict[str, Any]:
        """Make authenticated request to Grok API"""
        try:
            headers = {
                'Authorization': f'Bearer {self.api_key}',
                'Content-Type': 'application/json'
            }
            
            url = f"{self.api_base}/{endpoint}"
            response = requests.post(url, headers=headers, json=data, timeout=60)
            
            if response.status_code != 200:
                raise Exception(f"Grok API error: {response.status_code} - {response.text}")
            
            return response.json()
            
        except requests.exceptions.Timeout:
            raise Exception("Grok API request timed out")
        except requests.exceptions.RequestException as e:
            raise Exception(f"Grok API request failed: {e}")
        except Exception as e:
            logger.error(f"Grok API error: {e}")
            raise
    
    def chat_completion(self, messages: List[Dict[str, str]], 
                       temperature: Optional[float] = None,
                       max_tokens: Optional[int] = None) -> str:
        """Generate chat completion using Grok"""
        try:
            data = {
                'model': self.model,
                'messages': messages,
                'temperature': temperature or self.temperature,
                'max_tokens': max_tokens or self.max_tokens,
                'stream': False
            }
            
            response = self._make_request('chat/completions', data)
            
            # Extract the response content
            if 'choices' in response and len(response['choices']) > 0:
                content = response['choices'][0].get('message', {}).get('content', '')
                
                # Log token usage if available
                if 'usage' in response:
                    usage = response['usage']
                    logger.info(f"Grok API usage - prompt: {usage.get('prompt_tokens', 0)}, "
                              f"completion: {usage.get('completion_tokens', 0)}, "
                              f"total: {usage.get('total_tokens', 0)}")
                
                return content.strip()
            else:
                raise Exception("No response content from Grok API")
                
        except Exception as e:
            logger.error(f"Grok chat completion failed: {e}")
            raise
    
    def analyze_financial_data(self, data_context: str, user_query: str) -> str:
        """Analyze financial data with Grok's reasoning capabilities"""
        try:
            system_prompt = """You are Grok, an advanced AI financial analyst specializing in QuickBooks Online data analysis. 
            
            Your expertise includes:
            - Financial statement analysis and insights
            - Revenue forecasting and trend analysis  
            - Cash flow optimization recommendations
            - Business performance metrics interpretation
            - Actionable financial advice for small-mid businesses
            
            Guidelines for responses:
            - Be concise yet comprehensive in your analysis
            - Focus on actionable insights and recommendations
            - Use clear, business-friendly language
            - Highlight key trends, patterns, and anomalies
            - Provide specific next steps when possible
            - Be honest about limitations of the data
            
            When analyzing QuickBooks data:
            - Account types: Asset, Liability, Equity, Income, Expense
            - Key metrics: Revenue growth, profit margins, cash flow, expense ratios
            - Important dates and periods for trend analysis
            - Context matters - consider seasonality and business type
            """
            
            messages = [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": f"""
                Financial Data Context:
                {data_context}
                
                User Question: {user_query}
                
                Please provide a thorough analysis addressing the user's question with specific insights, trends, and actionable recommendations based on the QuickBooks data provided.
                """}
            ]
            
            return self.chat_completion(messages, temperature=0.3)  # Lower temperature for financial analysis
            
        except Exception as e:
            logger.error(f"Financial analysis failed: {e}")
            raise
    
    def generate_forecast(self, historical_data: List[Dict], 
                         forecast_type: str = "revenue", 
                         periods: int = 12) -> str:
        """Generate financial forecasts using Grok's reasoning"""
        try:
            data_summary = self._summarize_historical_data(historical_data, forecast_type)
            
            forecast_prompt = f"""
            Based on the following historical {forecast_type} data from QuickBooks Online, 
            generate a {periods}-period forecast with reasoning:
            
            Historical Data Summary:
            {data_summary}
            
            Please provide:
            1. Forecast values for the next {periods} periods
            2. Confidence level and reasoning
            3. Key assumptions made
            4. Risk factors to consider
            5. Recommendations for improving forecast accuracy
            
            Format the forecast in a clear, actionable way for business decision-making.
            """
            
            messages = [
                {"role": "system", "content": "You are Grok, an expert financial forecasting AI. Provide data-driven forecasts with clear reasoning and actionable insights."},
                {"role": "user", "content": forecast_prompt}
            ]
            
            return self.chat_completion(messages, temperature=0.2)  # Very low temperature for forecasting
            
        except Exception as e:
            logger.error(f"Forecast generation failed: {e}")
            raise
    
    def _summarize_historical_data(self, data: List[Dict], data_type: str) -> str:
        """Summarize historical data for analysis"""
        if not data:
            return f"No historical {data_type} data available"
        
        try:
            summary_lines = [f"Historical {data_type.title()} Data ({len(data)} records):"]
            
            # Add data summary based on type
            if data_type == "revenue":
                for item in data[-12:]:  # Last 12 periods
                    period = item.get('period', 'Unknown')
                    amount = item.get('amount', 0)
                    summary_lines.append(f"- {period}: ${amount:,.2f}")
            
            elif data_type == "accounts":
                account_summary = {}
                for account in data:
                    account_type = account.get('AccountType', 'Unknown')
                    balance = float(account.get('CurrentBalance', 0))
                    if account_type not in account_summary:
                        account_summary[account_type] = []
                    account_summary[account_type].append(balance)
                
                for acc_type, balances in account_summary.items():
                    total = sum(balances)
                    count = len(balances)
                    summary_lines.append(f"- {acc_type}: {count} accounts, Total: ${total:,.2f}")
            
            return "\n".join(summary_lines)
            
        except Exception as e:
            logger.warning(f"Data summarization failed: {e}")
            return f"Historical {data_type} data available ({len(data)} records)"
    
    def explain_financial_concept(self, concept: str, context: str = "") -> str:
        """Explain financial concepts in simple terms"""
        try:
            messages = [
                {"role": "system", "content": "You are Grok, an expert financial educator. Explain complex financial concepts in simple, actionable terms for small business owners."},
                {"role": "user", "content": f"""
                Please explain the financial concept: "{concept}"
                
                {f"Business context: {context}" if context else ""}
                
                Provide:
                1. Simple definition
                2. Why it matters for business
                3. How to improve/optimize it
                4. Common mistakes to avoid
                
                Keep it practical and actionable.
                """}
            ]
            
            return self.chat_completion(messages)
            
        except Exception as e:
            logger.error(f"Concept explanation failed: {e}")
            raise
    
    def health_check(self) -> Dict[str, Any]:
        """Check if Grok API is accessible"""
        try:
            # Simple test request
            messages = [{"role": "user", "content": "Hello, are you working?"}]
            response = self.chat_completion(messages, max_tokens=50)
            
            return {
                "status": "healthy",
                "model": self.model,
                "api_base": self.api_base,
                "response_length": len(response),
                "timestamp": datetime.now().isoformat()
            }
            
        except Exception as e:
            return {
                "status": "unhealthy",
                "error": str(e),
                "timestamp": datetime.now().isoformat()
            }

File: frontend/src/services/grokService.ts
/**
 * Grok API Integration Service for FinHelm.ai
 * Provides real-time preview functionality for custom agents
 */

import { AIModel } from '../types/agent';

export interface GrokPreviewRequest {
  prompt: string;
  query: string;
  model: AIModel;
  temperature: number;
  maxTokens: number;
  organizationId?: string;
  userId?: string;
}

export interface GrokPreviewResponse {
  response: string;
  executionTime: number;
  tokensUsed: number;
  cost?: number;
  model: string;
  success: boolean;
  error?: string;
}

export interface GrokConfig {
  apiKey?: string;
  baseUrl?: string;
  timeout?: number;
}

class GrokService {
  private config: GrokConfig;
  private debounceTimers = new Map<string, NodeJS.Timeout>();

  constructor(config: GrokConfig = {}) {
    this.config = {
      apiKey: config.apiKey || process.env.GROK_API_KEY,
      baseUrl: config.baseUrl || 'https://api.x.ai/v1',
      timeout: config.timeout || 30000,
    };
  }

  /**
   * Preview agent response with debouncing to avoid excessive API calls
   */
  async previewAgentResponse(
    request: GrokPreviewRequest,
    debounceMs: number = 1000
  ): Promise<GrokPreviewResponse> {
    const debounceKey = `${request.organizationId}-${request.userId}`;

    // Clear existing debounce timer
    const existingTimer = this.debounceTimers.get(debounceKey);
    if (existingTimer) {
      clearTimeout(existingTimer);
    }

    return new Promise((resolve, reject) => {
      const timer = setTimeout(async () => {
        this.debounceTimers.delete(debounceKey);
        
        try {
          const response = await this.callGrokAPI(request);
          resolve(response);
        } catch (error) {
          reject(error);
        }
      }, debounceMs);

      this.debounceTimers.set(debounceKey, timer);
    });
  }

  /**
   * Make direct call to Grok API without debouncing
   */
  async callGrokAPI(request: GrokPreviewRequest): Promise<GrokPreviewResponse> {
    const startTime = Date.now();

    try {
      // Check if we're in development/preview mode - use mock responses
      if (!this.config.apiKey || process.env.NODE_ENV === 'development') {
        return this.getMockGrokResponse(request, startTime);
      }

      // Prepare the system prompt and user message
      const systemPrompt = this.buildSystemPrompt(request.prompt);
      const userMessage = request.query;

      // Make API call to Grok
      const response = await fetch(`${this.config.baseUrl}/chat/completions`, {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${this.config.apiKey}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          model: this.mapModelName(request.model),
          messages: [
            {
              role: 'system',
              content: systemPrompt
            },
            {
              role: 'user',
              content: userMessage
            }
          ],
          temperature: request.temperature,
          max_tokens: request.maxTokens,
          stream: false,
        }),
        signal: AbortSignal.timeout(this.config.timeout || 30000),
      });

      if (!response.ok) {
        throw new Error(`Grok API error: ${response.status} ${response.statusText}`);
      }

      const data = await response.json();
      const executionTime = Date.now() - startTime;

      // Extract response and usage information
      const grokResponse = data.choices?.[0]?.message?.content || 'No response generated';
      const tokensUsed = data.usage?.total_tokens || 0;
      const cost = this.calculateCost(tokensUsed, request.model);

      return {
        response: grokResponse,
        executionTime,
        tokensUsed,
        cost,
        model: request.model,
        success: true,
      };

    } catch (error) {
      const executionTime = Date.now() - startTime;
      
      // Handle timeout errors
      if (error instanceof Error && error.name === 'TimeoutError') {
        return {
          response: '',
          executionTime,
          tokensUsed: 0,
          model: request.model,
          success: false,
          error: 'Request timed out. The AI model took too long to respond.',
        };
      }

      // Handle network and API errors
      return {
        response: '',
        executionTime,
        tokensUsed: 0,
        model: request.model,
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error occurred',
      };
    }
  }

  /**
   * Build a comprehensive system prompt for the agent
   */
  private buildSystemPrompt(customPrompt: string): string {
    const basePrompt = `You are an AI financial assistant integrated with FinHelm.ai, a comprehensive ERP co-pilot system.

Context:
- You have access to financial data from ERP systems (QuickBooks, Sage Intacct, NetSuite, Xero)
- Your responses should be professional, accurate, and actionable
- Focus on providing insights that help with financial decision-making
- Always consider compliance and audit requirements in your recommendations

Custom Instructions:
${customPrompt}

Response Guidelines:
- Provide specific, actionable insights
- Use financial terminology appropriately
- Include relevant data points when possible
- Suggest next steps or recommendations
- Format responses clearly with headers and bullet points when appropriate`;

    return basePrompt;
  }

  /**
   * Map internal model names to Grok API model names
   */
  private mapModelName(model: AIModel): string {
    const modelMap: Record<AIModel, string> = {
      'grok-beta': 'grok-beta',
      'gpt-4': 'gpt-4',
      'gpt-4-turbo': 'gpt-4-turbo-preview',
      'claude-3-opus': 'claude-3-opus',
      'claude-3-sonnet': 'claude-3-sonnet',
      'claude-3-haiku': 'claude-3-haiku',
    };

    return modelMap[model] || 'grok-beta';
  }

  /**
   * Calculate estimated cost based on token usage and model
   */
  private calculateCost(tokensUsed: number, model: AIModel): number {
    // Pricing per 1000 tokens (estimated)
    const pricingMap: Record<AIModel, number> = {
      'grok-beta': 0.002,
      'gpt-4': 0.03,
      'gpt-4-turbo': 0.01,
      'claude-3-opus': 0.015,
      'claude-3-sonnet': 0.003,
      'claude-3-haiku': 0.00025,
    };

    const pricePerToken = (pricingMap[model] || 0.002) / 1000;
    return tokensUsed * pricePerToken;
  }

  /**
   * Generate mock response for development/testing
   */
  private async getMockGrokResponse(
    request: GrokPreviewRequest, 
    startTime: number
  ): Promise<GrokPreviewResponse> {
    // Simulate network delay
    const delay = 1000 + Math.random() * 2000; // 1-3 seconds
    await new Promise(resolve => setTimeout(resolve, delay));

    const mockResponses = this.generateMockResponse(request.query, request.model);
    const executionTime = Date.now() - startTime;
    const estimatedTokens = Math.ceil(mockResponses.length / 4);

    return {
      response: mockResponses,
      executionTime,
      tokensUsed: estimatedTokens,
      cost: this.calculateCost(estimatedTokens, request.model),
      model: request.model,
      success: true,
    };
  }

  /**
   * Generate contextual mock responses based on query type
   */
  private generateMockResponse(query: string, model: AIModel): string {
    const queryLower = query.toLowerCase();
    
    if (queryLower.includes('rewrite') && queryLower.includes('email')) {
      return `**Professional Email Rewrite**

Subject: Monthly Expense Report - November 2024

Dear Management Team,

I hope this message finds you well. Please find attached our comprehensive expense report for November 2024.

**Executive Summary:**
• Total Expenses: $45,230
• Budget Variance: -3.2% (under budget)
• Key Categories: Travel (42%), Software (28%), Operations (30%)

**Notable Items:**
- Conference attendance generated valuable networking opportunities
- New software licenses aligned with digital transformation goals  
- Operational efficiency improvements reduced overhead by 8%

The detailed breakdown includes all supporting documentation and has been reviewed for accuracy and compliance with company policies.

Please let me know if you need any additional information or have questions about specific line items.

Best regards,
Finance Team

*Powered by ${model} | FinHelm.ai Custom Agent*`;
    }

    if (queryLower.includes('categorize') && queryLower.includes('transaction')) {
      return `**Transaction Categorization Analysis**

**Transaction Details:**
- Amount: $2,500.00
- Vendor: ABC Software Solutions
- Date: Current Period

**Recommended Categorization:**
✅ **Primary Category:** Software & Technology
✅ **Sub-Category:** Business Software Licenses  
✅ **Confidence Level:** 96%

**Analysis Factors:**
• Vendor name pattern matches software providers
• Amount consistent with annual software licensing
• Historical data shows similar transactions from this vendor
• Industry classification aligns with technology services

**Chart of Accounts:**
- **Account Code:** 6200 - Technology Expenses
- **Department:** IT Operations
- **Tax Treatment:** Deductible business expense

**Recommendations:**
1. Set up recurring categorization rule for this vendor
2. Review software inventory for license optimization
3. Consider multi-year agreements for cost savings

*Powered by ${model} | FinHelm.ai Custom Agent*`;
    }

    if (queryLower.includes('variance') || queryLower.includes('analysis')) {
      return `**Financial Variance Analysis**

**Period:** Current Month vs Budget

**Key Findings:**
📈 **Revenue Performance:** +8.3% above forecast
📊 **Expense Control:** -2.1% under budget  
💰 **Net Impact:** $12,450 favorable variance

**Revenue Analysis:**
• Client payments accelerated due to improved invoicing process
• New customer acquisition exceeded projections by 15%
• Recurring revenue streams showing strong stability

**Expense Highlights:**
• Marketing spend optimization saved $3,200
• Energy costs reduced through efficiency initiatives  
• Travel expenses normalized after Q2 spike

**Action Items:**
1. **Investigate:** Unusual increase in office supplies (+45%)
2. **Optimize:** Review vendor contracts expiring this quarter
3. **Monitor:** Customer concentration risk in top 3 accounts

**Forecast Impact:**
Based on current trends, we project continued favorable variances for the remainder of the fiscal year.

*Powered by ${model} | FinHelm.ai Custom Agent*`;
    }

    // Default response for general queries
    return `**Financial Intelligence Response**

**Query Analysis:** "${query}"

Based on your custom agent configuration and financial data context, here's my analysis:

**Key Insights:**
• **Data Integration:** Leveraging real-time ERP data for accurate analysis
• **Pattern Recognition:** Identifying trends across multiple financial dimensions  
• **Risk Assessment:** Evaluating potential compliance and operational risks
• **Actionable Recommendations:** Providing specific next steps for optimization

**Financial Context:**
Your organization's financial data shows healthy performance metrics with opportunities for optimization in several key areas. The integrated approach allows for comprehensive analysis across accounts payable, receivable, and general ledger functions.

**Recommendations:**
1. **Process Automation:** Implement workflow rules for similar scenarios
2. **Data Quality:** Ensure consistent categorization and coding standards
3. **Performance Monitoring:** Set up alerts for unusual patterns or variances
4. **Compliance Review:** Regular audit trails and documentation updates

**Next Steps:**
This analysis can be expanded with specific parameters, date ranges, or account filters to provide more targeted insights for your business needs.

*Powered by ${model} | FinHelm.ai Custom Agent*

*Note: This is a preview response. Production agents will access your actual financial data for more precise analysis.*`;
  }

  /**
   * Check API health and rate limits
   */
  async healthCheck(): Promise<{ status: 'healthy' | 'degraded' | 'down'; rateLimitRemaining?: number }> {
    try {
      // Simple health check - try to make a minimal API call
      const response = await fetch(`${this.config.baseUrl}/models`, {
        method: 'GET',
        headers: {
          'Authorization': `Bearer ${this.config.apiKey}`,
        },
        signal: AbortSignal.timeout(5000),
      });

      const rateLimitRemaining = response.headers.get('x-ratelimit-remaining');
      
      if (response.ok) {
        return {
          status: 'healthy',
          rateLimitRemaining: rateLimitRemaining ? parseInt(rateLimitRemaining) : undefined,
        };
      } else if (response.status === 429) {
        return { status: 'degraded', rateLimitRemaining: 0 };
      } else {
        return { status: 'down' };
      }
    } catch {
      return { status: 'down' };
    }
  }

  /**
   * Clean up debounce timers
   */
  dispose(): void {
    this.debounceTimers.forEach(timer => clearTimeout(timer));
    this.debounceTimers.clear();
  }
}

// Export singleton instance
export const grokService = new GrokService();
export default grokService;